### ISSUE TITLE
Composite objective formulation

### ISSUE THREAD

**Issue by whoburg** [convexengineering/gpkit#225]
I'm not convinced that the current composite objective formulation sweeps the weights over a convex set. We should revisit.

Discussion copied from e-mail thread:

---

PS square is an inf norm ball.

On Mar 30, 2015, at 4:51 PM, Warren W Hoburg wrote:

Nope, the boundary of the 1-norm ball is a diamond in 2D — so it’s the line connecting (0, 1) to (1, 0).

For nonnegative (w1, w2), it’s simply the set w1 + w2 = 1.

In higher dimensions the boundary (for nonnegative w) is a plane (thus a convex set).

Make sense?

On Mar 30, 2015, at 4:36 PM, Edward Burnell wrote:

Heh, that's true. I was thinking it could be interesting to specify the constraints in GP.

But in the 2D case, doesn't the "1-norm ball" (i.e. square) actually miss all the interesting values? it just checks each objective while the other is off, then both objectives on, but doesn't trade off between the objectives. This is another reason to prefer the 2-norm version...

Sorry, my descriptions were all messed up. Let me try again. Examples of convex sets (to pick the weights from) might be:
- the set of nonnegative reals (weights all nonnegative)
- the nonnegative subset of an inf norm ball (all weights between 0 and 1)
- the nonnegative subset of a 1-norm ball (weights are nonnegative and sum of all weights must be <= 1)
- the nonnegative subset of a 2-norm ball (weights are nonnegative and Euclidean norm of weights must be <= 1)
- the nonnegative _boundary_ of a 1-norm ball (weights are nonnegative and sum of weights must == 1)

The last one is the only one that has N-1 degrees of freedom instead of N. It’s also still a convex set!

Note that requiring the 2-norm to be exactly 1 (boundary of the 2-norm ball) would result in a non-convex set. That might be fine, but for reasons I don’t fully understand I think it’s best for the weights to come from a convex set.

Sorry for the confusion, but I hope the above makes sense.

-Woody

On Mar 30, 2015, at 9:11 AM, Warren W Hoburg wrote:

Those should all say “nonnegative portion” or “nonnegative subset” instead of "nonnegative half"
On Mar 30, 2015, at 9:09 AM, Warren W Hoburg wrote:

Figured I should just include everyone on this

On Mar 30, 2015, at 9:09 AM, Warren W Hoburg wrote:

Yeah… what I’m familiar with (though I admit I don’t know a lot about this) is a composite objective like the one described by Boyd — \sum_i lambda_i f_i(x). Then the (positive) lambda_i would usually come from a convex set. Examples of convex sets might be:
- the set of nonnegative reals (weights can be anything nonnegative)
- the nonnegative half of an inf norm ball (e.g. all weights are between 0 and 1, with no other restrictions)
- the nonnegative half of a 1 norm ball (e.g. weights are nonnegative and must sum to exactly one)
- the nonnegative half of a 2 norm ball (e.g. weights are nonnegative and must have euclidean norm equal to 1)

The last two options have only N-1 degrees of freedom (instead of N), which Ned pointed out is a nice feature.

It’s unclear to me how the current `composite_objective` fits into or relates to the above, but I feel like I might just be missing something simple.

-Woody

On Mar 30, 2015, at 12:42 AM, Cody Karcher wrote:

This looks wrong to me.  Ned, we may have to hash this out offline...

Basically the idea is to use linear weights to sweep out the Pareto Frontier.  It's common practice in MOO.  If you want a specific reference, I'm sure I can find one somewhere...

On Sun, Mar 29, 2015 at 9:17 PM, Warren W Hoburg wrote:
Cody,

Can you tell me where this objective formulation comes from?

Woody

On Mar 29, 2015, at 8:06 PM, Edward Burnell wrote:

It's from Cody! I used it as the default formulation because it reduces the number of weights you need from n to n-1.

On Mar 29, 2015, at 6:40 PM, Warren W Hoburg wrote:

Awesome Ned!

The formulation you use below (with the v_i = 1-w_i) is pretty specific in the way it couples the control of the weights together (as opposed to just letting them all range over a convex set). Can you briefly explain where it came from? (did you make it up, did you find it somewhere?)

Woody

On Mar 27, 2015, at 7:19 PM, Edward Burnell wrote:

Composite Objectives

Given <chart.png> posynomial objectives <chart.png>, you can sweep out the problem’s Pareto frontier with the composite objective:

<chart.png>

where <chart.png> and <chart.png>

GPkit has the helper function composite_objective for constructing these.

Usage

``` python
import numpy as np
import gpkit
gpkit.disableUnits()

L, W = gpkit.Variable("L"), gpkit.Variable("W")

eqns = [L >= 1, W >= 1, L*W == 10]

co_sweep = [0]+np.logspace(-6, 0, 10).tolist()

obj = gpkit.composite_objective(L+W, W**-1 * L**-3,
                                normsub={L:10, W: 10},
                                sweep=co_sweep)

gp = gpkit.GP(obj, eqns)
gp.solve()
```

The normsub argument allows specifies an expected value for your solution to normalize the different <chart.png> (you can also do this by hand). The feasibility of the problem should not depend on the normalization, but the spacing of the sweep will.

The sweep argument specifies what points between 0 and 1 you wish to sample the weights at. If you want different resolutions or spacings for different weights, the sweeps argument accepts a list of sweep arrays.

**Comment by bqpd** [comment]
replaced by https://github.com/convexengineering/gpkit/issues/1164
